pallets/creditcoin/src/lib.rs

```diff
@@ -619,13 +629,26 @@ pub mod pallet {
 		}
 
 		/// Registers an external address on `blockchain` and `network` with value `address`
-		#[pallet::weight(<T as Config>::WeightInfo::register_address(blockchain.as_bytes().len().unique_saturated_into(),address.as_slice().len().unique_saturated_into()))]
+		#[pallet::weight(<T as Config>::WeightInfo::register_address())]
 		pub fn register_address(
 			origin: OriginFor<T>,
 			blockchain: Blockchain,
 			address: ExternalAddress,
+			signature: sp_core::ecdsa::Signature,
 		) -> DispatchResult {
 			let who = ensure_signed(origin)?;
+
+			let message = sp_io::hashing::sha2_256(who.encode().as_slice());
+			let message = &sp_io::hashing::blake2_256(message.as_ref());
+			let signature = <[u8; 65]>::from(signature);
+			let raw_pkey = secp256k1_ecdsa_recover_compressed(&signature, message)
+				.map_err(|_| Error::<T>::EcdsaRecoveryFailed)?;
+			let recreated_address = helpers::external_address_generator(&blockchain, &address)
+				.ok_or(Error::<T>::AddressFormatNotSupported)?(
+				&sp_core::ecdsa::Public::from_raw(raw_pkey)
+			);
```

Comment:
It feels weird to return a function object from `external_address_generator`, only to immediately invoke the returned function. Why not remove the indirection? Instead something like
```rust
pub fn generate_external_address(
	blockchain: &Blockchain,
	reference: &ExternalAddress,
    public_key: &Public,
) -> Option<ExternalAddress>
```
feels more straightforward. You also avoid allocating the `Box` for the returned function that way.

Reply:
Decoupling selecting an address generator from generating it. It may become increasingly expensive to select the method, especially with bitcoin and Other. Maybe it is not that important in this particular case but if you were to generate a  few addresses the overhead may quickly stack. Also, I don't like dragging parameters across functions, Public is not needed to select an address generator, it makes the call stack shorter and the function more specific. Again, it may look trivial until we start supporting more address protocols. I will benchmark if the extra boxing and indirection is worth the cost though.

Reply:
Benchmarking register_adress shows an average difference within 1%. I don't think it would matter performance wise.

Reply:
We chose to stick with the straightforward one. Marking it resolved.